godot-core/src/builtin/meta/godot_compat/mod.rs

```diff
@@ -0,0 +1,120 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ */
+
+mod impls;
+
+use godot_ffi as sys;
+
+use crate::builtin::{Variant, VariantConversionError};
+
+use super::{GodotFfiVariant, GodotType};
+
+/// Indicates that a type has some canonical Godot type that can represent it.
+///
+/// The type specified here is what will be used to pass this type across to ffi-boundary to/from Godot.
+/// Generally [`ToGodot`] needs to be implemented to pass a type to Godot, and [`FromGodot`] to receive this
+/// type from Godot.
+pub trait GodotCompatible {
+    /// The type used for ffi-passing.
+    type Via: GodotType;
+}
+
+/// Defines the canonical conversion to Godot for a type.
+///
+/// It is assumed that all the methods return equal values given equal inputs. Additionally it is assumed
+/// that if [`FromGodot`] is implemented, converting to Godot and back again will return a value equal to the
+/// starting value.
+///
+/// Violating these assumptions is safe but will give unexpected results.
+pub trait ToGodot: Sized + GodotCompatible {
```

Comment:
Can `GodotCompatible` itself be unsized? Or should the `Sized` bound be moved one level up?

Reply:
I guess it might be possible to impl `GodotCompatible` for an unsized type, and create a custom getter/setter if you wanna be able to export unsized types? It'd be a very niche use-case though, but it doesn't cost much to support it. It just means we need to have `Sized` as bounds on `To/FromGodot` instead of on `GodotCompatible`.

Reply:
I see, let's keep it then for now!